src/metadata/metadata_passive_io.c

```diff
@@ -0,0 +1,377 @@
+/*
+ * Copyright(c) 2012-2021 Intel Corporation
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include "ocf/ocf.h"
+
+#include "metadata.h"
+#include "metadata_passive_io.h"
+#include "metadata_collision.h"
+#include "metadata_segment_id.h"
+#include "metadata_internal.h"
+#include "metadata_io.h"
+#include "metadata_raw.h"
+#include "metadata_segment.h"
+#include "../concurrency/ocf_concurrency.h"
+#include "../ocf_def_priv.h"
+#include "../ocf_priv.h"
+#include "../utils/utils_cache_line.h"
+#include "../utils/utils_io.h"
+#include "../utils/utils_pipeline.h"
+#include "../concurrency/ocf_pio_concurrency.h"
+#include "../engine/engine_common.h"
+
+#define MAX_PASSIVE_IO_SIZE (32*MiB)
+
+
+static inline void _reset_cline(ocf_cache_t cache, ocf_cache_line_t cline)
+{
+	/* The cacheline used to be dirty but it is not anymore so it needs to be
+	   moved to a clean lru list */
+	ocf_lru_clean_cline(cache, &cache->user_parts[PARTITION_DEFAULT].part,
+			cline);
+
+	ocf_lru_rm_cline(cache, cline);
+	ocf_metadata_set_partition_id(cache, cline, PARTITION_FREELIST);
+}
+
+static inline void remove_from_freelist(ocf_cache_t cache,
+		ocf_cache_line_t cline)
+{
+	ocf_part_id_t lru_list;
+	struct ocf_lru_list *list;
+
+	lru_list = (cline % OCF_NUM_LRU_LISTS);
+	list = ocf_lru_get_list(&cache->free, lru_list, true);
+
+	OCF_METADATA_LRU_WR_LOCK(cline);
+	ocf_lru_remove_locked(cache, list, cline);
+	OCF_METADATA_LRU_WR_UNLOCK(cline);
+}
+
+static inline void remove_from_default(ocf_cache_t cache,
+		ocf_cache_line_t cline)
+{
+	ocf_part_id_t part_id = PARTITION_DEFAULT;
+	ocf_part_id_t lru_list;
+	struct ocf_lru_list *list;
+
+	lru_list = (cline % OCF_NUM_LRU_LISTS);
+	list = ocf_lru_get_list(&cache->user_parts[part_id].part, lru_list, false);
+
+	OCF_METADATA_LRU_WR_LOCK(cline);
+	ocf_lru_remove_locked(cache, list, cline);
+	OCF_METADATA_LRU_WR_UNLOCK(cline);
+
+	env_atomic_dec(&cache->user_parts[part_id].part.runtime->curr_size);
+
+}
+
+static void handle_previously_invalid(ocf_cache_t cache,
+		ocf_cache_line_t cline)
+{
+	ocf_core_id_t core_id;
+	uint64_t core_line;
+	uint32_t lock_idx = ocf_metadata_concurrency_next_idx(cache->mngt_queue);
+
+	ocf_metadata_get_core_info(cache, cline, &core_id, &core_line);
```

Comment:
Can you please add a comment here explaining, that cacheline mapping is fixed at the moment thanks to pio lock being held? This is another reason to have pio locks, let's try to document this somehow.

Reply:
... also status bits are fixed for the same reason (next line - metadata_test_dirty())